README.md
# doqr
doqr allows you to build node.js docker images without docker, allowing you to build a node.js docker image *from within a docker container*. This means you can build the container image on Kubernetes or Openshift - or locally in a docker container for more hermetic builds.

It will pull an image you specify from a given registry, add the node.js application from a given folder, and push the result to a(nother) given registry.

## How to install
```
npm install -g doqr
```

## How to make a Doqr binary
```
npm install; pkg cli.js
```

## How to use

This will pull the `node:13-slim` image from Docker hub, build the image by adding the application in `src/`, and push the result to the given registry, and set time timestamp of files in the created layers and configs to the current timestamp of the latest git commit.

```
doqr --fromImage node:13-slim --folder src/ --toImage myapp:latest --toRegistry https://registry.example.com/v2/ --setTimeStamp=$(git show -s --format="%aI" HEAD)
```

### Command line options 
```
Usage: doqr [options]

Options:
  --fromRegistry <registry url>  Optional: URL of registry to pull base image from - Default: https://registry-1.docker.io/v2/
  --fromImage <name:tag>         Required: Image name of base image - [path/]image:tag
  --fromToken <token>            Optional: Authentication token for from registry
  --toRegistry <registry url>    Optional: URL of registry to push base image to - Default: https://registry-1.docker.io/v2/
  --toImage <name:tag>           Required: Image name of target image - [path/]image:tag
  --toToken <token>              Optional: Authentication token for target registry
  --toTar <path>                 Optional: Export to tar file
  --registry <path>              Optional: Convenience argument for setting both from and to registry
  --token <path>                 Optional: Convenience argument for setting token for both from and to registry
  --folder <full path>           Required: Base folder of application - Default: .
  --user <user>                  Optional: User account to run process in container - default: 1001
  --workdir <directory>          Optional: Workdir where app will be added and run from - default: /opt/app-root/src
  --entrypoint <entrypoint>      Optional: Entrypoint when starting container
  --cmd <command>                Optional: CMD for docker image
  --setTimeStamp <timestamp>     Optional: Set a specific ISO 8601 timestamp on all entries (e.g. git commit hash). Default: 1970 in tar files, and now on manifest/config
  --verbose                      Verbose logging
  --allowInsecureRegistries      Allow insecure registries (with self-signed/untrusted cert)
  --layerOwner   <gid:uid>       Optional: Set specific gid and uid on files in the added layers
  -h, --help                     output usage information
```

CHANGELOG.md
# Changelog

## [0.4.2] - 2020-08-10

### Modified
- Improved scoping of folder flag

### Removed
- Custom content feature

## [0.4.1] - 2020-08-08

### Modified
- Reverted change of 0.4.0 - must find a more clever way

## [0.4.0] - 2020-08-08

### Modified
- Do not download base-image layers if src and dest registries are the same 

## [0.3.5] - 2020-08-08

### Modified
- Modified app layer creation

## [0.3.4] - 2020-08-08

### Modified
- Attempt to transfer symlinks to app layer content

## [0.3.3] - 2020-08-08

### Modified
- Change file permissions in generated app layer
- Minor cleanup

## [0.3.2] - 2020-08-07

### Modified
- Minor cleanup

## [0.3.1] - 2020-08-07

### Added
- Added vercel/pkg

## [0.3.0] - 2020-08-07

### Added
- Support for setting image CMD

### Modified
- Merged app and dependency layers

## [0.2.0] - 2020-08-05

### Added
- Support for setting the owner of the work folder (gid:uid)

## [0.1.0] - 2020-03-30

### Added
- Support custom layer (drops default entrypoint, user, workdir, node_modules layer and app layer). Only adds the specified files

## [0.0.11] - 2020-03-30

### Modified
- Don't include .git and .gitignore if in same folder

## [0.0.10] - 2020-03-08

### Modified
- Updated the README with improved example and missing option for the timestamp

## [0.0.9] - 2020-03-08

### Removed
- Removed files from npm package and simplified package.json to use defaults

## [0.0.8] - 2020-03-08

### Added
- Ability to set a specific timestamp on all files/tars/configs to support hermetic builds. Typically one would use the git commit time (`--setTimeStamp=$(git show -s --format="%aI" HEAD)`). If omitted, the timestamp is set to epoch 0. [`420e248`](https://github.com/eoftedal/doqr/commit/420e248e4daf5470e91834f11a52633a566f5783)


tarExporter.js
const tar = require('tar');
const fs = require('fs').promises;
const fse = require('fs-extra');
const path = require('path');

const logger = require('./logger');

const tarDefaultConfig = {
  preservePaths: false, 
  portable: true, 
  follow: true  
};

async function saveToTar(fromdir, tmpdir, toPath, repoTags, options) {
  logger.info('Creating ' +  toPath + ' ...');

  let manifestFile = path.join(fromdir, 'manifest.json');
  let manifest = await fse.readJson(manifestFile);
  let configFile = path.join(fromdir, manifest.config.digest.split(':')[1] + '.json');
  let config = await fse.readJson(configFile);

  let tardir = path.join(tmpdir, 'totar'); 
  await fs.mkdir(tardir);
  let layers = await Promise.all(
    manifest.layers
      .map(x => x.digest.split(':')[1])
      .map(async x => {
        let fn = x + ((await fse.pathExists(path.join(fromdir, x + '.tar.gz'))) ? '.tar.gz' : '.tar');
        await fse.copy(path.join(fromdir, fn), path.join(tardir, fn));
        return fn;
      })
  );

  let simpleManifest = [{
    config  : 'config.json',
    repoTags: [repoTags],
    layers  : layers
  }];
  await fs.writeFile(path.join(tardir, 'manifest.json'), JSON.stringify(simpleManifest));
  await fs.writeFile(path.join(tardir, 'config.json'), JSON.stringify(config));
  await tar.c(Object.assign({}, tarDefaultConfig, {
    cwd: tardir,
    file: toPath,
    noMtime: (!options.setTimeStamp),
    mtime: options.setTimeStamp
  }), ['config.json', 'manifest.json'].concat(layers));
  logger.info('Finished ' + toPath);
}

module.exports = {
  saveToTar : saveToTar
};

image/registry.js
const https = require('https');
const http = require('http');
const URL = require('url');
const fs = require('fs').promises;
const path = require('path');
const fse = require('fs-extra');

const fileutil = require('../utils/fileutil');
const logger = require('../utils/logger');

const redirectCodes = [307,303,302];

function request(options, callback) {
  return (options.protocol == 'https:' ? https : http).request(options, res => {
    callback(res);
  });
}

function isOk(httpStatus) {
  return httpStatus >= 200 && httpStatus < 300;
}

function getHash(digest) {
  return digest.split(':')[1];
}

function parseImage(imageStr) {
  let ar = imageStr.split(':');
  let tag = ar[1] || 'latest';
  let ipath = ar[0];
  return { path: ipath, tag: tag };
}

function toError(res) {
  return `Unexpected HTTP status ${res.statusCode} : ${res.statusMessage}`;
}

function dl(uri, headers) {
  logger.debug('dl', uri);
  return new Promise((resolve, reject) => {
    followRedirects(uri, headers, (err, res) => {
      if (err) return reject(err);
      logger.debug(res.statusCode, res.statusMessage, res.headers['content-type'], res.headers['content-length']);
      if (!isOk(res.statusCode)) return reject(toError(res));
      let data = [];
      res
        .on('data', chunk => data.push(chunk.toString()))
        .on('end', () => {
          resolve(data.reduce((a,b) => a.concat(b)));
        });  
    });
  });
}

async function dlJson(uri, headers) {
  var data = await dl(uri, headers);
  return JSON.parse(Buffer.from(data).toString('utf-8'));
}

function dlToFile(uri, file, headers) {
  return new Promise((resolve, reject) => {
    followRedirects(uri, headers, (err, res) => {
      if (err) return reject(err);
      logger.debug(res.statusCode, res.statusMessage, res.headers['content-type'], res.headers['content-length']);
      if (!isOk(res.statusCode)) return reject(toError(res));
      res.pipe(require('fs').createWriteStream(file))
        .on('finish', () => {
          logger.debug('Done ' + file + ' - ' + res.headers['content-length'] + ' bytes ');
          resolve();
        });  
    });
  });  
}

function followRedirects(uri, headers, cb, count=0) {
  logger.debug('rc', uri);
  let options = URL.parse(uri);
  options.headers = headers;
  options.method = 'GET';
  request(options, res => {
    if (redirectCodes.includes(res.statusCode)) {
      if (count > 10) return cb('Too many redirects for ' + uri); 
      return followRedirects(res.headers.location, headers, cb, count+1);
    }
    cb(null, res);
  }).end();
}

function buildHeaders(accept, auth) {
  let headers = { accept: accept };
  if (auth) headers.authorization = auth;
  return headers;
}

function headOk(url, headers) {
  return new Promise((resolve, reject) => {
    logger.debug(`HEAD ${url}`);
    let options = URL.parse(url);
    options.headers = headers;
    options.method = 'HEAD';
    request(options, res => {
      logger.debug(`HEAD ${url}`, res.statusCode);
      if (res.statusCode == 404) return resolve(false);
      if (res.statusCode == 200) return resolve(true);
      reject(toError(res));
    }).end();
  });  
}

function uploadContent(uploadUrl, file, fileConfig, auth) {
  return new Promise((resolve, reject) => {
    logger.debug('Uploading: ', file);
    let url = uploadUrl;
    if (fileConfig.digest) url += (url.indexOf('?') == -1 ? '?' : '&') + 'digest=' + fileConfig.digest;
    let options = URL.parse(url);
    options.method = 'PUT';
    options.headers = {
      authorization    : auth,
      'content-length' : fileConfig.size,
      'content-type'   : fileConfig.mediaType
    };
    logger.debug('POST', url);
    let req = request(options, res => {
      logger.debug(res.statusCode, res.statusMessage, res.headers['content-type'], res.headers['content-length']);
      if ([200,201,202,203].includes(res.statusCode)) {
        resolve();
      } else {
        var data = [];
        res.on('data', d => data.push(d.toString()));
        res.on('end', () => {
          reject(`Error uploading to ${uploadUrl}. Got ${res.statusCode} ${res.statusMessage}:\n${data.join('')}`);
        });
      }
    });
    require('fs').createReadStream(file).pipe(req);
  });  
}

function Registry(registryBaseUrl, token) {
  const auth = 'Bearer ' + token;

  async function exists(image, layer) {
    let url = `${registryBaseUrl}${image.path}/blobs/${layer.digest}`;
    return await headOk(url, buildHeaders(layer.mediaType, auth));
  }

  async function uploadLayerContent(uploadUrl, layer, dir) {
    logger.info(layer.digest);
    let file = path.join(dir, getHash(layer.digest) + (layer.mediaType.includes('tar.gzip')  ? '.tar.gz' : '.tar'));
    await uploadContent(uploadUrl, file, layer, auth);
  }

  function getUploadUrl(image) {
    return new Promise((resolve, reject) => {
      let url = `${registryBaseUrl}${image.path}/blobs/uploads/`;
      let options = URL.parse(url);
      options.method = 'POST';
      options.headers = { authorization: auth };
      request(options, res => {
        logger.debug('POST', `${url}`, res.statusCode);
        if (res.statusCode == 202) {
          resolve(res.headers['location']);        
        } else {
          var data = [];
          res.on('data', c => data.push(c.toString()))
            .on('end', () => {
              reject(`Error getting upload URL from ${url}. Got ${res.statusCode} ${res.statusMessage}:\n${data.join('')}`);
            });
        }
      }).end();
    });
  }

  async function dlManifest(image) {
    return await dlJson(
      `${registryBaseUrl}${image.path}/manifests/${image.tag}`, 
      buildHeaders('application/vnd.docker.distribution.manifest.v2+json', auth)
    );
  }

  async function dlConfig(image, config) {
    return await dlJson(
      `${registryBaseUrl}${image.path}/blobs/${config.digest}`,
      buildHeaders('*/*', auth)
    );
  }

  async function dlLayer(image, layer, folder) {
    logger.info(layer.digest);
    let file = getHash(layer.digest) + (layer.mediaType.includes('tar.gzip')  ? '.tar.gz' : '.tar');
    await dlToFile(
      `${registryBaseUrl}${image.path}/blobs/${layer.digest}`,
      path.join(folder, file),
      buildHeaders(layer.mediaType, auth)
    );
    return file;
  }

  async function upload(imageStr, folder) {
    let image = parseImage(imageStr);
    let manifestFile = path.join(folder, 'manifest.json');
    let manifest = await fse.readJson(manifestFile);

    logger.info('Checking layer status...');
    let layerStatus = await Promise.all(manifest.layers.map(async l => {
      return { layer: l, exists: (await exists(image, l)) };
    }));
    let layersForUpload = layerStatus.filter(l => !l.exists);
    logger.debug('Needs upload:', layersForUpload.map(l => l.layer.digest));

    logger.info('Uploading layers...');
    await Promise.all(layersForUpload.map(async l => {
      let url = await getUploadUrl(image);
      await uploadLayerContent(url, l.layer, folder);
    }));

    logger.info('Uploading config...');
    let configUploadUrl = await getUploadUrl(image);
    let configFile = path.join(folder, getHash(manifest.config.digest) + '.json');
    await uploadContent(configUploadUrl, configFile, manifest.config, auth);

    logger.info('Uploading manifest...');
    let manifestSize = await fileutil.sizeOf(manifestFile);
    await uploadContent(
      `${registryBaseUrl}${image.path}/manifests/${image.tag}`, 
      manifestFile, 
      { mediaType: manifest.mediaType, size: manifestSize }, 
      auth
    );
  }


  async function download(imageStr, folder) {
    let image = parseImage(imageStr);

    logger.info('Downloading manifest...');
    let manifest = await dlManifest(image);
    await fs.writeFile(path.join(folder, 'manifest.json'), JSON.stringify(manifest));

    logger.info('Downloading config...');
    let config = await dlConfig(image, manifest.config);
    await fs.writeFile(path.join(folder, 'config.json'), JSON.stringify(config));

    logger.info('Downloading layers...');
    await Promise.all(manifest.layers.map(layer => dlLayer(image, layer, folder)));

    logger.info('Image downloaded.');
  }

  return {
    download: download,
    upload: upload
  };
}

function DockerRegistry(auth) {
  const registryBaseUrl = 'https://registry-1.docker.io/v2/';

  async function getToken(image) {
    let resp = await dlJson(`https://auth.docker.io/token?service=registry.docker.io&scope=repository:${image.path}:pull`);
    return resp.token;
  }
  async function download(imageStr, folder) {
    let image = parseImage(imageStr);
    if (!auth) auth = await getToken(image);
    await Registry(registryBaseUrl, auth).download(imageStr, folder);
  }

  async function upload(imageStr, folder) {
    if (!auth) throw new Error('Need auth token to upload to Docker');
    await Registry(registryBaseUrl, auth).upload(imageStr, folder);
  }

  return {
    download: download,
    upload: upload
  };
}

module.exports = {
  Registry : Registry,
  DockerRegistry: DockerRegistry
};

image/appLayerCreator.js
const tar = require('tar');
const fs = require('fs').promises;
const fse = require('fs-extra');
const fss = require('fs');
const path = require('path');
const crypto = require('crypto');
const Gunzip = require('minizlib').Gunzip;

const fileutil = require('../utils/fileutil');
const logger = require('../utils/logger');

const ignore = ['.git', '.gitignore', '.npmrc', '.DS_Store', 'npm-debug.log', '.svn', '.hg', 'CVS']

function statCache(layerOwner) {
  if (!layerOwner) return null;
  // We use the stat cache to overwrite uid and gid in image. 
  // A bit hacky
  const statCacheMap = new Map();
  const a = layerOwner.split(":");
  const gid = parseInt(a[0]);
  const uid = parseInt(a[1]);
  return {
    get: function(name) {
      if (statCacheMap.has(name)) return statCacheMap.get(name);
      let stat = fss.statSync(name);
      stat.uid = uid;
      stat.gid = gid;
      stat.atime = new Date(0);
      stat.mtime = new Date(0);
      stat.ctime = new Date(0);
      stat.birthtime = new Date(0);
      stat.atimeMs = 0;
      stat.mtimeMs = 0;
      stat.ctimeMs = 0;
      stat.birthtimeMs = 0;
      statCacheMap.set(name, stat);
      return stat;
    },
    set: function(name, stat) {
      statCacheMap.set(name, stat);
    },
    has: function(name) {
      return true;
    }
  };
}

function calculateHashOfBuffer(buf) {
  let hash = crypto.createHash('sha256');
  hash.update(buf);
  return hash.digest('hex');
}

function calculateHash(path) {
  return new Promise((resolve, reject) => {
    let hash = crypto.createHash('sha256');
    let stream = require('fs').createReadStream(path);
    stream.on('error', err => reject(err));
    stream.on('data', chunk => hash.update(chunk));
    stream.on('end', () => resolve(hash.digest('hex')));
  });
}

function addEmptyLayer(config, options, operation, action) {
  logger.info(`Applying ${operation}`);
  config.history.push({
    created: options.setTimeStamp || new Date().toISOString(),
    created_by: '/bin/sh -c #(nop) ' + operation,
    empty_layer: true
  });
  action(config);
}

async function getHashOfUncompressed(file) {
  return new Promise((resolve, reject) => {
    let hash = crypto.createHash('sha256');
    let gunzip = new Gunzip();
    gunzip.on('data', chunk => hash.update(chunk));
    gunzip.on('end', () => resolve(hash.digest('hex')));
    gunzip.on('error', err => reject(err));
    require('fs').createReadStream(file).pipe(gunzip).on('error', err => reject(err));
  });
}

async function addDataLayer(tmpdir, todir, options, config, layers, files, comment) {
  logger.info('Adding layer for ' + comment + ' ...');
  let layerFile = path.join(todir, 'layer.tar.gz');
  if (options.layerOwner) logger.info("Setting file ownership to: " + options.layerOwner);
  let contentFiles = fss.readdirSync(options.folder);
  await tar.c(Object.assign({},{
    statCache: statCache(options.layerOwner),
    portable: !options.layerOwner,
    prefix: options.workdir,
    cwd: options.folder,
    file: layerFile,
    gzip: true,
    noMtime: (!options.setTimeStamp),
    mtime: options.setTimeStamp
  }),contentFiles);
  let fhash = await calculateHash(layerFile);
  let finalName = path.join(todir, fhash + '.tar.gz');
  await fse.move(layerFile, finalName);
  layers.push({
    mediaType : 'application/vnd.docker.image.rootfs.diff.tar.gzip',
    size: await fileutil.sizeOf(finalName),
    digest: 'sha256:' + fhash
  });
  let dhash = await getHashOfUncompressed(finalName);
  config.rootfs.diff_ids.push('sha256:' + dhash);
  config.history.push({
    created: options.setTimeStamp || new Date().toISOString(),
    created_by: 'doqr',
    comment: comment
  });
}

async function copyLayers(fromdir, todir, layers) {
  await Promise.all(layers.map(async layer => {
    let file = layer.digest.split(':')[1] + (layer.mediaType.includes('tar.gzip')  ? '.tar.gz' : '.tar');
    await fse.copy(path.join(fromdir, file), path.join(todir, file));
  }));
}

function parseCommandLineToParts(entrypoint) {
  return entrypoint.split('"')
      .map((p,i) => {
        if (i % 2 == 1) return [p];
        return p.split(' ');
      })
      .reduce((a, b) => a.concat(b), [])
      .filter(a => a != '');
}

function splitLabelsIntoObject(labelsString) {
  let labels = {};
  labelsString.split(',').map(l => l.split('=')).map(l => labels[l[0]] = l[1]);
  return labels;
}

async function addAppLayers(options, config, todir, manifest, tmpdir) {
  addEmptyLayer(config, options, `WORKDIR ${options.workdir}`, config => config.config.WorkingDir = options.workdir);
  if(options.entrypoint){
    let entrypoint = parseCommandLineToParts(options.entrypoint);
    addEmptyLayer(config, options, `ENTRYPOINT ${JSON.stringify(entrypoint)}`, config => config.config.Entrypoint = entrypoint);
  }
  if(options.cmd){
    let command = parseCommandLineToParts(options.cmd);
    addEmptyLayer(config, options, `CMD ${JSON.stringify(command)}`, config => config.config.Cmd = command);
  }
  addEmptyLayer(config, options, `USER ${options.user}`, config => {
    config.config.user = options.user;
    config.container_config.user = options.user;
  });
  addLabelsLayer(options, config, todir, manifest, tmpdir)
  let appFiles = (await fs.readdir(options.folder)).filter(l => !ignore.includes(l));

  await addDataLayer(tmpdir, todir, options, config, manifest.layers, appFiles, 'app');
}

async function addLabelsLayer(options, config, todir, manifest, tmpdir) {
  if (options.labels) {
    let labels = splitLabelsIntoObject(options.labels);
    addEmptyLayer(config, options, `LABELS ${options.labels}`, config => {
      config.config.labels = labels;
      config.container_config.labels = labels;
    });
  }
}

async function addLayers(tmpdir, fromdir, todir, options) {

  logger.info('Parsing image ...');
  let manifest = await fse.readJson(path.join(fromdir, 'manifest.json'));
  let config = await fse.readJson(path.join(fromdir, 'config.json'));

  logger.info('Adding new layers...');
  await copyLayers(fromdir, todir, manifest.layers);
  await addAppLayers(options, config, todir, manifest, tmpdir);

  logger.info('Writing final image...');
  let configContent = Buffer.from(JSON.stringify(config));
  let configHash = calculateHashOfBuffer(configContent);
  let configFile = path.join(todir, configHash + '.json');
  await fs.writeFile(configFile, configContent);
  manifest.config.digest = 'sha256:' + configHash;
  manifest.config.size = await fileutil.sizeOf(configFile);
  await fs.writeFile(path.join(todir, 'manifest.json'), JSON.stringify(manifest));
}

module.exports = {
  addLayers : addLayers
};



cli.js
#!/usr/bin/env node

const os = require('os');
const program  = require('commander');
const path = require('path');
const fse = require('fs-extra');
const fs = require('fs');

const Registry = require('./image/registry').Registry;
const DockerRegistry = require('./image/registry').DockerRegistry;
const appLayerCreator = require('./image/appLayerCreator');
const fileutil = require('./utils/fileutil');

const logger = require('./utils/logger');

const possibleArgs = {
    '--fromImage <name:tag>'        : 'Required: Image name of base image - [path/]image:tag',
    '--toImage <name:tag>'          : 'Required: Image name of target image - [path/]image:tag',
    '--folder <full path>'          : 'Required: Base folder of application - Default: .',
    '--fromRegistry <registry url>' : 'Optional: URL of registry to pull base image from - Default: https://registry-1.docker.io/v2/',
    '--fromToken <token>'           : 'Optional: Authentication token for from registry',
    '--toRegistry <registry url>'   : 'Optional: URL of registry to push base image to - Default: https://registry-1.docker.io/v2/',
    '--toToken <token>'             : 'Optional: Authentication token for target registry',
    '--registry <path>'             : 'Optional: Convenience argument for setting both from and to registry',
    '--token <path>'                : 'Optional: Convenience argument for setting token for both from and to registry',
    '--user <user>'                 : 'Optional: User account to run process in container - default: 1001',
    '--workdir <directory>'         : 'Optional: Workdir where app will be added and run from - default: /opt/app-root/src',
    '--cmd <command>'               : 'Optional: CMD for docker image',
    '--entrypoint <entrypoint>'     : 'Optional: Entrypoint when starting container',
    '--labels <labels>'             : 'Optional: Comma-separated list of key value pairs to use as labels',
    '--setTimeStamp <timestamp>'    : 'Optional: Set a specific ISO 8601 timestamp on all entries (e.g. git commit hash). Default: 1970 in tar files, and now on manifest/config',
    '--verbose'                     : 'Verbose logging',
    '--allowInsecureRegistries'     : 'Allow insecure registries (with self-signed/untrusted cert)',
    '--layerOwner <gid:uid>'        : 'Optional: Set specific gid and uid on files in the added layers',
};

const keys = Object.keys(possibleArgs)
    .map(k => k.split(' ')[0].replace('--', ''));

Object.keys(possibleArgs)
    .reduce((program, k) => program.option(k, possibleArgs[k]), program)
    .parse(process.argv);

let options = {
    folder: '.',
    workdir : '/opt/app-root/src',
    user: '1001'
};

keys.map(k => options[k] = program[k] || options[k]);

function exitWithErrorIf(check, error) {
    if (check) {
        logger.error('ERROR: ' + error);
        program.outputHelp();
        process.exit(1);
    }
}

if (options.verbose) logger.enableDebug();
if (options.allowInsecureRegistries) process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = 0;

exitWithErrorIf(options.registry && options.fromRegistry, 'Do not set both --registry and --fromRegistry');
exitWithErrorIf(options.registry && options.toRegistry, 'Do not set both --registry and --toRegistry');
exitWithErrorIf(options.token && options.fromToken, 'Do not set both --token and --fromToken');
exitWithErrorIf(options.token && options.toToken, 'Do not set both --token and --toToken');

if (options.setTimeStamp) {
    try {
        options.setTimeStamp = new Date(options.setTimeStamp);
    } catch(e) {
        exitWithErrorIf(true, 'Failed to parse date: ' + e);
    }
    logger.info('Setting all dates to: ' + options.setTimeStamp);
}
if (options.layerOwner) {
    if (!options.layerOwner.match("^[0-9]+:[0-9]+$")) {
        exitWithErrorIf(true, 'layerOwner should be on format <number>:<number> (e.g. 1000:1000) but was: ' + options.layerOwner);
    }
}
if (options.registry) {
    options.fromRegistry = options.registry;
    options.toRegistry = options.registry;
}
if (options.token) {
    options.fromToken = options.token;
    options.toToken = options.token;
}

exitWithErrorIf(!options.folder, '--folder must be specified');
exitWithErrorIf(!options.fromImage, '--fromImage must be specified');
exitWithErrorIf(!options.toImage, '--toImage must be specified');
exitWithErrorIf(!options.toRegistry && !options.toTar, 'Must specify either --toTar or --toRegistry');
exitWithErrorIf(!options.toRegistry && !options.toToken && !options.toTar, 'A token must be given when uploading to docker hub');

if(options.toRegistry && options.toRegistry.substr(-1) != '/') options.toRegistry += '/';
if(options.fromRegistry && options.fromRegistry.substr(-1) != '/') options.fromRegistry += '/';

if (!options.fromRegistry && !options.fromImage.split(':')[0].includes('/')) {
    options.fromImage = 'library/' + options.fromImage;
}

async function run(options) {
    if (!(await fse.pathExists(options.folder))) throw new Error('No such folder: ' + options.folder);

    const tmpdir = require('fs').mkdtempSync(path.join(os.tmpdir(),'doqr-'));
    logger.debug('Using ' + tmpdir);
    let fromdir = await fileutil.ensureEmptyDir(path.join(tmpdir, 'from'));
    let todir = await fileutil.ensureEmptyDir(path.join(tmpdir, 'to'));

    let fromRegistry = options.fromRegistry ? new Registry(options.fromRegistry, options.fromToken) : new DockerRegistry(options.fromToken);
    let toRegistry = new Registry(options.toRegistry, options.toToken);

    await fromRegistry.download(options.fromImage, fromdir);
    await appLayerCreator.addLayers(tmpdir, fromdir, todir, options);
    await toRegistry.upload(options.toImage, todir);

    logger.debug('Deleting ' + tmpdir + ' ...');
    await fse.remove(tmpdir);
    logger.debug('Done');
}

run(options).then(() => {
    logger.info('Done!');
    process.exit(0);
}).catch(error => {
    logger.error(error);
    process.exit(1);
});